Added support for requests that contain an ETA datetime in the future

weetster opened the pull request:
Hello,

I'm using celery-batches for a project that requires failed tasks to be retried at a later time (for example, if a web API request fails). I documented the approach that I was hoping to use in the case of failure and retry (`apply_async(countdown=10)`) in this pull request, which was inspired by the discussion in issue #10. I noticed that the approach did not work, because celery-batches@0.6 does not handle Celery Request objects that have an `.eta` field. I've added support to celery-batches for handling this scenario, and I hope that you will consider merging my changes into the main branch for inclusion in an upcoming 0.7 release.

Thank you.

clokep commented:
Thanks! Will try to take a look soon! 👍

clokep commented:
Looks like there's a couple of lint errors, could you run `flake8` (or take a look at the CI output) and fix those up? Thanks!

clokep commented on docs/index.rst:
```diff
@@ -93,6 +93,23 @@ Using the API is done as follows::
 
+.. note::
+
+    The use of ``current_app.backend.mark_as_retry()`` and ``current_app.backend.mark_as_failure()`` are not currently supported.
```
`mark_as_failure` is used in the unit test, but is mentioned here as not supported -- can you elaborate on what this means then?

weetster replied:
You're right, I think I had left `mark_as_failure()` in there during testing. It can be removed, and the new unit tests still pass, so I'll remove it.

clokep replied:
I think this comment is still outstanding?

weetster replied:
Addressed in this commit: https://github.com/clokep/celery-batches/pull/59/commits/08a7a1bad637fb3e0b62386cdc3a7a7c7f7ee0b3

clokep commented on celery_batches/__init__.py:
```diff
@@ -34,11 +35,21 @@ def consume_queue(queue):
 
     """
     get = queue.get_nowait
+    requests_to_requeue = []
```
I suspect that handling of the `eta` would be better to position in the `Strategy` (which seems to be where it is in celery itself, see https://github.com/celery/celery/blob/fbda0089f08d7f2a8f00925dbc0b6e10bd779251/celery/worker/strategy.py#L179-L185

But maybe that's not possible?

weetster replied:
In `celery`, it appears the Strategy's `task_message_handler()` is responsible for scheduling the ETA request by using the `call_at()` function, which exists in some sort of timer object. I assume this timer object keeps track of the request and handles it at the appropriate time.

For `celery-batches` to do the same thing in its Strategy, it would need some similar method of tracking the request until the ETA time has passed. It seems the only way `celery-batches` can track requests over time is via its Queue. I could move the ETA logic to the Strategy, but I'm not sure what I would do with the request if its ETA time hasn't passed yet and it needs to be stored somewhere that isn't the Queue (since the Queue would only be for requests that are ready).

I think it makes sense to keep the ETA logic in `consume_queue()` and use the Queue as both temporary storage for requests that aren't ready, and temporary storage for requests that are ready. But I'm fairly new to both Celery and celery-batches, so I'm open to changing it if you think it makes sense.

clokep replied:
I think storing them in the `Queue` is fine, I guess I'm trying to avoid adding Celery-specific logic into `consume_queue`, which is a rather dumb helper right now.

I think the `Strategy` could implement the same thing, pretty much:

1. Consume the queue.
2. For each `Request`, check it is ready, if not put it back in the queue.
3. Call the batch task.

Either implementation pretty much has the same logic, I think? There is a bad failure mode either way though, if you have more than `flush_every` in your queue that have not yet reached their ETA than batches would just keep trying to fire the tasks, but actually have nothing to do.

Maybe it makes sense to do something more like...

1. Consume the queue + the pending requests queue.
2. For each `Request`, check it is ready, if not put in a pending requests queue.
3. Call the batch task (if there are any requests left).

That still has some oddities, but might be more OK? This means that a task with an ETA on it would fire within `flush_interval` of the ETA -- we should probably document how ETA + flushing works regardless.

What do you think? Is that reasonable to your application?

weetster replied:
I've pushed a commit that I think makes sense, and all unit tests passed for me. I've also manually tested my project that uses my celery-batches repo, and everything seems good there too.

clokep commented on celery_batches/__init__.py:
```diff
@@ -34,11 +35,21 @@ def consume_queue(queue):
         try:
-            yield get()
+            req = get()
+            if req.eta is not None:
```
Do we need to handle `req.countdown` as well as `req.eta` or does something, somewhere convert `countdown` into `eta`?

Seems this is done for us:

* v1 task: https://github.com/celery/celery/blob/fbda0089f08d7f2a8f00925dbc0b6e10bd779251/celery/app/amqp.py#L381-L384
* v2 task: https://github.com/celery/celery/blob/fbda0089f08d7f2a8f00925dbc0b6e10bd779251/celery/app/amqp.py#L294-L300

clokep replied:
(To be clear, there's nothing to do here -- I answered my own question!)

weetster replied:
Good to point out. I think both countdown and eta will end up as `req.eta` in the end, by the time the request is handled in celery-batches.

clokep replied:
Yep, I think this is fine! I was just curious if it handled one and not the other, but it seems the celery code does it already! 🎉

weetster commented:
Hey, thanks for the code review. I just noticed it now, since it was posted the same day that I left for a week-long vacation. I'll take a look at addressing the comments as soon as I can.

clokep commented:
> Hey, thanks for the code review. I just noticed it now, since it was posted the same day that I left for a week-long vacation. I'll take a look at addressing the comments as soon as I can.

No problem! Do let me know if you have any questions or disagree with any of the comments. I think I also bitrotted this PR decently by adding type hints, sorry about that. 😢 Hopefully they make it a bit easier to see what's going on though!

clokep commented:
I need to document this, but you should be able to do the following to run the lint checks:

```bash
$ pip install -r requirements/pkgutils.txt
$ flake8
$ isort
$ black
$ mypy
```

clokep commented on celery_batches/__init__.py:
```diff
@@ -285,15 +287,36 @@ def apply(
+                    ready_requests.append(req)
+                else:
+                    # ETA has not elapsed, add to pending queue
+                    pending_requests.append(req)
```
I don't think `pending_requests` is needed, we've already emptied `self._pending` on line 296, so I think this can be:

```suggestion
                    self._pending.put(req)
```

Which means we don't need to re-iterate over `pending_requests` later on.

clokep commented on t/integration/tasks.py:
```diff
@@ -41,3 +41,20 @@ def cumadd(requests: List[SimpleRequest]) -> None:
+            # Odd, success
+            current_app.backend.mark_as_done(request.id, True, request=request)
+        else:
+            # Even, so modify to be odd next time around and retry
+            request.args = (request.args[0] + 1,)
+            retry_if_even.apply_async(
+                args=request.args, kwargs=request.kwargs, countdown=3
+            )
```
It feels weird/unnecessary to modify the incoming `SimpleRequest` object although there isn't really anything wrong with it.

```suggestion
            # Even, modify it to be odd next time around and retry.
            retry_if_even.apply_async(
                args=(request.args[0] + 1,), kwargs=request.kwargs, countdown=3
            )
```

clokep commented on t/integration/tasks.py:
```diff
@@ -41,3 +41,20 @@ def cumadd(requests: List[SimpleRequest]) -> None:
         current_app.backend.mark_as_done(request.id, result, request=request)
+
+
+@shared_task(base=Batches, flush_every=2, flush_interval=1)
```
I made this much shorter in #60, so let's match that.

```suggestion
@shared_task(base=Batches, flush_every=2, flush_interval=0.1)
```

clokep commented on t/integration/test_batches.py:
```diff
@@ -270,3 +270,30 @@ def acks(requests: List[SimpleRequest]) -> None:
+    result_1 = retry_if_even.delay(1)
+    retry_if_even.delay(2)
+
+    # The flush interval is 1 second and the retry interval is 3 seconds,
+    # this is longer.
+    sleep(5)
```
To match our other changes:

```suggestion
    # The flush interval is 0.1 seconds and the retry interval is 0.3 seconds,
    # this is longer.
    sleep(0.5)
```

weetster replied:
I had to mess with the timing a bit to get it to be consistently successful, but it's shorter than it was before.

clokep commented on t/integration/test_batches.py:
```diff
@@ -270,3 +270,30 @@ def acks(requests: List[SimpleRequest]) -> None:
+
+    assert result_1.get() is True
+
+    # The retried task is not accessible here for validation.
```
I'm not really sure what you mean -- what would we be validating?

weetster replied:
I can't do the equivalent of

```
result_2 = retry_if_even.delay(2)
assert result_2.get() is True
```

because that task fails, and an entirely new task is created inside retry_if_even() that is not accessible from this unit test.

clokep replied:
That makes me a bit concerned that this solution isn't really tenable for retries in the general case. It breaks the contract that the task ID is usable to get results.

I think supporting the countdown / ETA is still a necessarily requirement, but maybe we can figure out how to properly support it in a follow-up?

clokep replied:
I'm wondering if we can simplify the test significantly. Maybe something like:

1. Call task A
2. Call task B with a countdown of X seconds (longer than the flush interval)
3. Wait for the worker to work.
4. Assert that A processed.
5. Sleep such that flush interval is long enough to process B.
6. Wait for the worker to work.
7. Assert that B was processed.

I think this means we wouldn't even need a new task in the tests?

weetster replied:
I realized that when re-scheduling a task to retry, the task_id can be specified in apply_async() such that you can still call get() on the original result and the contract is not broken. I've updated the tests and the docs to show this. I would prefer to leave the tests as-is rather than change them as you suggested, since the tests in their current state exercise both the ETA functionality and the retry functionality. Curious to hear your thoughts though.

weetster replied:
I got the idea for re-using the task_id from looking at how celery itself handles retries internally, which is very similar to what I implemented:

The Task retry() function will create a signature from the original request:
https://github.com/celery/celery/blob/master/celery/app/task.py#L706

Which includes the same task ID:
https://github.com/celery/celery/blob/0a3487b882398ef44d3be34a57b8354e986511af/celery/app/task.py#L607
https://github.com/celery/celery/blob/0a3487b882398ef44d3be34a57b8354e986511af/celery/app/task.py#L115

And then the retry() function will call apply_async() on that signature:
https://github.com/celery/celery/blob/master/celery/app/task.py#L733

As for the mark_as_retry() function, I was mistaken by saying that it's not supported. I think it was working as intended, but its only purpose is to set the state of a task as RETRY in the backend for tracking purposes. If mark_as_retry() is not used, the task stays in the PENDING state, which also works. The other missing piece of the puzzle was the ETA and task_id support.

I thought all this made sense, and I can add test cases for pure ETA requests in addition to the retry test cases.

Would this approach still be acceptable?

clokep replied:
Thanks for the cross-reference. I'll take another look.

clokep replied:
> I thought all this made sense, and I can add test cases for pure ETA requests in addition to the retry test cases.

I think the retry test case doesn't really belong here though -- it isn't testing celery-batches, it is testing application code. I think it is clearer to just call `apply_async()` with a countdown and with an `eta` in tests directly.

Looking at the `Task.retry()` function I wonder if we should wrap some of that logic up into a nice method though. I think that could make a nice follow-up to give the incoming requests an easy `retry()` method?

weetster replied:
Thanks for the second review. That's a good point, the unit test relies more on application behaviour, which we can't guarantee/control in the outside world.

clokep replied:
I made this change in 69522d25b03d726fdcb912c361b84e713f55c3b2, by the way. I think it is a bit clearer.

clokep commented on celery_batches/__init__.py:
```diff
@@ -285,15 +287,36 @@ def apply(
+        for req in pending_requests:
+            self._pending.put(req)
+
+        if self._buffer.qsize() == 0 and self._pending.qsize() == 0:
```
I think `_buffer` is always going to be empty here?

```suggestion
        if not ready_requests and self._pending.qsize() == 0:
```

weetster commented:
I was just making fixes for the workflow, didn't see your new comments until now. Thanks, will take a look!

weetster commented:
All set for another look. Thanks again!

clokep commented:
@weetster This is back to you, note that if you can't get the tests to run with live brokers yet, we can skip it. I think there's something flakey with returning results in those cases. (You can see they skip another test also...)

weetster commented:
> @weetster This is back to you, note that if you can't get the tests to run with live brokers yet, we can skip it. I think there's something flakey with returning results in those cases. (You can see they skip another test also...)

Thanks, the tests work well on my local machine, so I added a skip similar to the other test case that you mentioned.

clokep commented on docs/index.rst:
```diff
@@ -93,6 +93,28 @@ Using the API is done as follows::
         from celery import current_app
         current_app.backend.mark_as_done(request.id, response, request=request)
 
+.. note::
+
+    In order to retry a failed task, the task must be re-executed with a matching ``task_id``,
+    as shown below::
+
+        @app.task(base=Batches, flush_every=100, flush_interval=10)
+        def wot_api(requests):
+            # Do things that might fail
+            # responses = [...]
+
+            for response, request in zip(reponses, requests):
+                if response is True:
+                    app.backend.mark_as_done(request.id, response, request=request)
+                else:
+                    # Retry the task with the same arguments and task_id, 10 seconds from now
+                    wot_api.apply_async(
+                        args=request.args,
+                        kwargs=request.kwargs,
+                        countdown=10,
+                        task_id=request.id,
+                    )
```
This page changed quite a bit...

```suggestion
Retrying tasks
##############

In order to retry a failed task, the task must be re-executed with a matching ``task_id``,
as shown below:

.. code-block:: python

    @app.task(base=Batches, flush_every=100, flush_interval=10)
    def my_task(requests):
        for request in requests:
            # Do things which might fail.
            result = ...
            if not result:
                # Retry the task with the same arguments and task_id, 10 seconds from now.
                my_task.apply_async(
                    args=request.args,
                    kwargs=request.kwargs,
                    countdown=10,
                    task_id=request.id,
                )
```

clokep commented:
If it is easier I can make the remaining changes and push them to your branch (I *think* I have the right permissions to do that...)

weetster commented:
> If it is easier I can make the remaining changes and push them to your branch (I _think_ I have the right permissions to do that...)

Sure, that's ok with me, if it's easier for you to make the changes to how you want it. Let me know if there's anything else I should do. :+1:

clokep commented:
@weetster I pushed a few commits to:

* Update the tests a bit like we discussed.
* Update the documentation (to be more inline with other changes on `main`).
* Properly handle timezones (mostly be copying the behavior inside of Celery's default strategy).

Does this all look reasonable to you? If you could give it a quick test and make sure it still works for you I would appreciate it! 👍 The current release already has a lot in it so would like to include this and then get it out the door. :shipit:

weetster commented:
@clokep Thanks for the extra commits, good to handle timezones properly. celery-batches unit tests pass for me, and my project that relies on celery-batches also passes its unit tests with the latest changes. Looks good to me for a release.

clokep commented:
Thanks for checking! 🎉

weetster commented:
You're welcome, and thanks for accepting the changes! They will be a huge help to me and hopefully others.
